script/copy_deduplicate

```diff
@@ -130,42 +167,88 @@ group.add_argument(
     ),
 )
 
+temporary_dataset = None
+
+
+def get_temporary_dataset(client):
+    """Get a cached reference to the dataset used for destinationless queries."""
+    global temporary_dataset
+    if temporary_dataset is None:
+        # look up the dataset used for query results without a destination
+        dry_run = bigquery.QueryJobConfig(dry_run=True)
+        destination = client.query("SELECT 1", dry_run).destination
+        temporary_dataset = client.dataset(destination.dataset_id, destination.project)
+    return temporary_dataset
+
+
+def get_temporary_table(client, date):
+    """Generate a random writable anonynmous table reference."""
+    return get_temporary_dataset(client).table(f"anon{uuid4().hex}${date:%Y%m%d}")
```

Comment:
It would be nice to avoid having this logic. Is it possible to let BQ assign temp tables, and then fetch the name from the query results? Or do we have to set an explicit table name because of the size of the results?

Also, do we know that BQ will automatically clean up tables in this dataset, even if we've created them explicitly rather than having BQ generate them?

Also, does this function follow the exact format of BQ's generated table names? If so, we should document that intention.

Reply:
> It would be nice to avoid having this logic. Is it possible to let BQ assign temp tables, and then fetch the name from the query results? Or do we have to set an explicit table name because of the size of the results?

we have to set an explicit table name because of result size, `time_partitioning`, and `clustering_fields`. result size is fine if we set `allow_large_results`, but `time_partitioning` and `clustering_fields` require `destination` to be set.

I also tried leaving out `time_partitioning` and `clustering_fields` while setting `dry_run`, then re-using that destination, but that returns the error message: `Cannot explicitly modify anonymous table`.

> Also, do we know that BQ will automatically clean up tables in this dataset, even if we've created them explicitly rather than having BQ generate them?

The temporary results dataset has a default expiration of 1 day, but I can confirm from my tests that it only impacts partition expiration, which is why i'll be adding explicit table expiration.

> Also, does this function follow the exact format of BQ's generated table names? If so, we should document that intention.

it follows it closely enough that the web console shows a link to `Temporary table` as the destination, but not closely enough that it would collide with a name generated by BigQuery. I will add an explanation in the docstring.